runtime/caml/major_gc.h

```diff
@@ -21,21 +21,28 @@
 #include "freelist.h"
 #include "misc.h"
 
+/* An interval of a single object to be scanned.
+   The end pointer must always be one-past-the-end of a heap block,
+   but the start pointer is not necessarily the start of the block */
+typedef struct {
+  value* start;
+  value* end;
+} mark_entry;
```

Comment:
I believe this is at least no worse than the current implementation.

(I think it might even be correct: if a truncated object is on the mark stack, then it has tag < No_scan_tag, which means truncate will overwrite all of its fields with Val_unit / header-with-odd-tag, so all of the to-be-marked fields become immediates)

Reply:
Ok, fair enough. (I am suspiciaous about the old implementation, because it continuously read the size from the head, and I wonder whether there is an assumption somewhere that this size is constant.).